src/_pytest/assertion/util.py

```diff
@@ -197,49 +201,51 @@ def _diff_text(left: str, right: str, verbose: int = 0) -> List[str]:
     explanation = []  # type: List[str]
 
     if verbose < 1:
-        i = 0  # just in case left or right has zero length
-        for i in range(min(len(left), len(right))):
-            if left[i] != right[i]:
+        i = 0  # just in case "result" or "expected" has zero length
+        for i in range(min(len(result), len(expected))):
+            if result[i] != expected[i]:
                 break
         if i > 42:
             i -= 10  # Provide some context
             explanation = [
                 "Skipping %s identical leading characters in diff, use -v to show" % i
             ]
-            left = left[i:]
-            right = right[i:]
-        if len(left) == len(right):
-            for i in range(len(left)):
-                if left[-i] != right[-i]:
+            result = result[i:]
+            expected = expected[i:]
+        if len(result) == len(expected):
+            for i in range(len(result)):
+                if result[-i] != expected[-i]:
                     break
             if i > 42:
                 i -= 10  # Provide some context
                 explanation += [
                     "Skipping {} identical trailing "
                     "characters in diff, use -v to show".format(i)
                 ]
-                left = left[:-i]
-                right = right[:-i]
+                result = result[:-i]
+                expected = expected[:-i]
     keepends = True
-    if left.isspace() or right.isspace():
-        left = repr(str(left))
-        right = repr(str(right))
+    if result.isspace() or expected.isspace():
+        result = repr(str(result))
+        expected = repr(str(expected))
         explanation += ["Strings contain only whitespace, escaping them using repr()"]
+    # "expected" is the base against which we compare "result",
+    # see https://github.com/pytest-dev/pytest/issues/3333
     explanation += [
         line.strip("\n")
-        for line in ndiff(left.splitlines(keepends), right.splitlines(keepends))
+        for line in ndiff(expected.splitlines(keepends), result.splitlines(keepends))
```

Comment:
Hmmm really sorry to comment this now, but I think there might have been a miscommunication here: in my mind we would only swap order of the parameters to `ndiff` like you did here (`ndiff(right, left)`), but keep using the terms "left" and "right" in the code.

I find `left` and `right` on the code unambiguous, the original problem as I see it was just the order of the parameters to `ndiff`, which produced a diff which was contrary to "result == expected" that most people consider obvious.

With that in mind, and if others agree, do you think it would be possible to revert the changes done to this file, except for the parameter swap done in `ndiff`?

Reply:
In my opinion, the ambiguity depends a lot on whether `assert result is expected` should really the way how people write their assertions or if it is just a suggestion.

I feel that in the first case, *result* and *expected* transport more meaning. Bu I aggree that *left* and *right* would be better if it is only a friendly suggestion to put the result on the left and the expectation on the right. It would then be no problem to revert my changes, of couse. :)

Reply:
I agree, but `assert result is expected` is kind of an implicit suggestion, while "left" vs "right" doesn't assume that.

> Bu I aggree that left and right would be better if it is only a friendly suggestion to put the result on the left and the expectation on the right

Yes, this is how I see it. We suggest users write assertions that way and when have to make a non-obvious choice (like the diff direction) we assume that's how assertions are written, but when there is an unambiguous way I think this should be preferred.

That's from my personal experience: I've never had doubt about what was being shown was presented as a "Left" vs "Right" POV, but the diff direction was always off for me. 😁

Reply:
I updated the PR